### PR TITLE
[DO NOT MERGE] Move parallel_reduce into a central location

### PR DESCRIPTION
**Pull request by cpuhrsch**
Make it easier to swap out a threading implementation by putting all calls to a single location.

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/Parallel.h [hunk 1]
```diff
@@ -51,4 +51,35 @@ void parallel_for(
...
 }
 
+template <class scalar_t, class F, class SF>
+inline scalar_t parallel_reduce(
+    int64_t begin,
+    int64_t end,
+    int64_t grain_size,
+    scalar_t ident,
+    F f,
+    SF sf) {
```
Comment (apaszke) [inline aten/src/ATen/Parallel.h]: Any chance for a better name than sf? It’s not clear at all what is that supposed to be

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp [hunk 2]
```diff
@@ -72,7 +67,7 @@ struct Reduction {
     }
     int64_t batch = numel / (n * stride);
     bool paralellize = batch * n > internal::TBB_GRAIN_SIZE;
-    parallel_for(batch, 1, paralellize, [=](int64_t b) {
+    _parallel_for(batch, 1, paralellize, [=](int64_t b) {
```
Comment (apaszke) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp]: Can we refactor this to use regular parallel_apply at some point?

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp [hunk 3]
```diff
@@ -23,19 +17,22 @@ static inline int64_t round_down(int64_t a, int64_t m) {
...
+    parallel_for(0, size / step, 1, [func, step](int64_t begin, int64_t end) {
+      int64_t k = begin * step;
+      for (int64_t i = begin; i < end; i++, k += step) {
+        func(k);
+      }
+    });
   } else {
-    for (int64_t i = 0; i != end; i += step) {
+    for (int64_t i = 0; i != size; i += step) {
       func(i);
```
Comment (apaszke) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp]: This function is really weird. Func is applied both in a blocked fashion (and without any epilogue), or in a scalar way? What is going on in here?!

**Review by apaszke — APPROVED** [review]

**Review by colesbury — APPROVED** [review]
I'm not convinced of the merits of moving all the calls to TBB through a few methods. Replacing TBB is not so likely that it justifies changing all the calls.

This TBB calls are well documented and the ordering of arguments is consistent (internally and with C++). For example, `parallel_for(start, end, step)` maps nicely to the common `for(int i = start; i < end; i+= step)`

The new wrappers make the code more complicated (e.g. `_parallel_for`). It's harder for someone new to the code to understand what it does.

**Comment by cpuhrsch** [comment]
@colesbury I agree with your sentiment, but at the same time we're seeing issues with collisions between threadpool implementations (tbb and openmp). It's good to be able to try different ones. Users might even want to use their own. If we don't do this now, it'll be much harder in the future. The api you described is easy to add. If you look at the [parallel_for.h](https://github.com/01org/tbb/blob/d31fcc68980efbe912dc8cec4978fbb77fb3cf9f/include/serial/tbb/parallel_for.h) header in tbb you'll see that a lot of the calls come from the need to support different partitioners. I'm aware that this changes when you also want to support multiple ranges.

**Comment by cpuhrsch** [comment]
Latest ASAN error

```
jenkins@a47375922d5c:~/workspace$ python test/test_torch.py TestTorch.test_cpu_parallel
.
----------------------------------------------------------------------
Ran 1 test in 0.523s

OK
=================================================================
==148==ERROR: AddressSanitizer: heap-use-after-free on address 0x6160005eff4a at pc 0x7fdac7609813 bp 0x7fda2c7a9d60 sp 0x7fda2c7a9d58
READ of size 1 at 0x6160005eff4a thread T39
==148==AddressSanitizer: while reporting a bug found another one. Ignoring.
==148==AddressSanitizer: while reporting a bug found another one. Ignoring.
==148==AddressSanitizer: while reporting a bug found another one. Ignoring.
==148==AddressSanitizer: while reporting a bug found another one. Ignoring.
==148==AddressSanitizer: while reporting a bug found another one. Ignoring.
==148==AddressSanitizer: while reporting a bug found another one. Ignoring.
==148==AddressSanitizer: while reporting a bug found another one. Ignoring.
==148==AddressSanitizer: while reporting a bug found another one. Ignoring.
==148==AddressSanitizer: while reporting a bug found another one. Ignoring.
==148==AddressSanitizer: while reporting a bug found another one. Ignoring.
==148==AddressSanitizer: while reporting a bug found another one. Ignoring.
==148==AddressSanitizer: while reporting a bug found another one. Ignoring.
==148==AddressSanitizer: while reporting a bug found another one. Ignoring.
==148==AddressSanitizer: while reporting a bug found another one. Ignoring.
==148==AddressSanitizer: while reporting a bug found another one. Ignoring.
==148==AddressSanitizer: while reporting a bug found another one. Ignoring.
==148==AddressSanitizer: while reporting a bug found another one. Ignoring.
    #0 0x7fdac7609812 in tbb::internal::generic_scheduler::is_worker() const /var/lib/jenkins/workspace/third_party/tbb/src/tbb/scheduler.h:571:26
    #1 0x7fdac7609bc8 in tbb::internal::generic_scheduler::master_outermost_level() const /var/lib/jenkins/workspace/third_party/tbb/src/tbb/scheduler.h:551:13
    #2 0x7fdac76094d7 in tbb::task_scheduler_init::internal_terminate(bool) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/governor.cpp:349:13
    #3 0x7fdac76095f6 in tbb::task_scheduler_init::terminate() /var/lib/jenkins/workspace/third_party/tbb/src/tbb/governor.cpp:359:5
    #4 0x7fdac5d92f70 in tbb::task_scheduler_init::~task_scheduler_init() /var/lib/jenkins/workspace/aten/src/ATen/../../../third_party/tbb/include/tbb/task_scheduler_init.h:132:13
    #5 0x7fdadb87ccce in (anonymous namespace)::run(void*) /opt/conda/conda-bld/compilers_linux-64_1520532893746/work/.build/src/gcc-7.2.0/libstdc++-v3/libsupc++/atexit_thread.cc:75
    #6 0x7fdadc80c438 in __nptl_deallocate_tsd.part.4 (/lib/x86_64-linux-gnu/libpthread.so.0+0x6438)
    #7 0x7fdadc80d86f in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x786f)
    #8 0x7fdadc54341c in clone /build/glibc-Cl5G7W/glibc-2.23/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:109

0x6160005eff4a is located 202 bytes inside of 520-byte region [0x6160005efe80,0x6160005f0088)
freed by thread T39 here:
    #0 0x7fdadcb2acc0 in __interceptor_cfree (/usr/lib/llvm-5.0/lib/clang/5.0.0/lib/linux/libclang_rt.asan-x86_64.so+0x107cc0)
    #1 0x7fdac7656a23 in tbb::internal::padded_free(void*) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/cache_aligned_allocator.cpp:226:9
    #2 0x7fdac7656c53 in tbb::internal::NFS_Free(void*) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/cache_aligned_allocator.cpp:205:5
    #3 0x7fdac761ea45 in tbb::internal::generic_scheduler::free_scheduler() /var/lib/jenkins/workspace/third_party/tbb/src/tbb/scheduler.cpp:295:9
    #4 0x7fdac762271e in tbb::internal::generic_scheduler::cleanup_worker(void*, bool) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/scheduler.cpp:1208:7
    #5 0x7fdac7612c00 in tbb::internal::market::cleanup(rml::job&) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/market.cpp:734:9
    #6 0x7fdac7617af8 in tbb::internal::rml::private_worker::run() /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:285:15
    #7 0x7fdac7617659 in tbb::internal::rml::private_worker::thread_routine(void*) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:223:11
    #8 0x7fdadc80d6b9 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x76b9)

previously allocated by thread T39 here:
    #0 0x7fdadcb2ae88 in malloc (/usr/lib/llvm-5.0/lib/clang/5.0.0/lib/linux/libclang_rt.asan-x86_64.so+0x107e88)
    #1 0x7fdac7656957 in tbb::internal::padded_allocate(unsigned long, unsigned long) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/cache_aligned_allocator.cpp:210:43
    #2 0x7fdac7656c09 in tbb::internal::NFS_Allocate(unsigned long, unsigned long, void*) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/cache_aligned_allocator.cpp:196:20
    #3 0x7fdac76231e5 in tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::allocate_scheduler(tbb::internal::market&) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/custom_scheduler.h:118:19
    #4 0x7fdac762762f in tbb::internal::allocate_scheduler(tbb::internal::market&) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/scheduler.cpp:42:12
    #5 0x7fdac7622078 in tbb::internal::generic_scheduler::create_worker(tbb::internal::market&, unsigned long) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/scheduler.cpp:1147:28
    #6 0x7fdac7612cd1 in tbb::internal::market::create_one_job() /var/lib/jenkins/workspace/third_party/tbb/src/tbb/market.cpp:747:28
    #7 0x7fdac7617853 in tbb::internal::rml::private_worker::run() /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:267:32
    #8 0x7fdac7617659 in tbb::internal::rml::private_worker::thread_routine(void*) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:223:11
    #9 0x7fdadc80d6b9 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x76b9)

Thread T39 created by T31 here:
    #0 0x7fdadca823cd in pthread_create (/usr/lib/llvm-5.0/lib/clang/5.0.0/lib/linux/libclang_rt.asan-x86_64.so+0x5f3cd)
    #1 0x7fdac761c494 in rml::internal::thread_monitor::launch(void* (*)(void*), void*, unsigned long) /var/lib/jenkins/workspace/aten/src/ATen/../../../third_party/tbb/src/rml/include/../server/thread_monitor.h:221:12
    #2 0x7fdac7619e7a in tbb::internal::rml::private_worker::wake_or_launch() /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:300:21
    #3 0x7fdac7618c2d in tbb::internal::rml::private_server::wake_some(int) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:394:17
    #4 0x7fdac7619497 in tbb::internal::rml::private_server::propagate_chain_reaction() /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:161:13
    #5 0x7fdac76177aa in tbb::internal::rml::private_worker::run() /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:261:15
    #6 0x7fdac7617659 in tbb::internal::rml::private_worker::thread_routine(void*) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:223:11
    #7 0x7fdadc80d6b9 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x76b9)

Thread T31 created by T28 here:
    #0 0x7fdadca823cd in pthread_create (/usr/lib/llvm-5.0/lib/clang/5.0.0/lib/linux/libclang_rt.asan-x86_64.so+0x5f3cd)
    #1 0x7fdac761c494 in rml::internal::thread_monitor::launch(void* (*)(void*), void*, unsigned long) /var/lib/jenkins/workspace/aten/src/ATen/../../../third_party/tbb/src/rml/include/../server/thread_monitor.h:221:12
    #2 0x7fdac7619e7a in tbb::internal::rml::private_worker::wake_or_launch() /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:300:21
    #3 0x7fdac7618c2d in tbb::internal::rml::private_server::wake_some(int) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:394:17
    #4 0x7fdac7619497 in tbb::internal::rml::private_server::propagate_chain_reaction() /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:161:13
    #5 0x7fdac76177aa in tbb::internal::rml::private_worker::run() /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:261:15
    #6 0x7fdac7617659 in tbb::internal::rml::private_worker::thread_routine(void*) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:223:11
    #7 0x7fdadc80d6b9 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x76b9)

Thread T28 created by T25 here:
    #0 0x7fdadca823cd in pthread_create (/usr/lib/llvm-5.0/lib/clang/5.0.0/lib/linux/libclang_rt.asan-x86_64.so+0x5f3cd)
    #1 0x7fdac761c494 in rml::internal::thread_monitor::launch(void* (*)(void*), void*, unsigned long) /var/lib/jenkins/workspace/aten/src/ATen/../../../third_party/tbb/src/rml/include/../server/thread_monitor.h:221:12
    #2 0x7fdac7619e7a in tbb::internal::rml::private_worker::wake_or_launch() /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:300:21
    #3 0x7fdac7618c2d in tbb::internal::rml::private_server::wake_some(int) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:394:17
    #4 0x7fdac7619497 in tbb::internal::rml::private_server::propagate_chain_reaction() /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:161:13
    #5 0x7fdac76177aa in tbb::internal::rml::private_worker::run() /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:261:15
    #6 0x7fdac7617659 in tbb::internal::rml::private_worker::thread_routine(void*) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:223:11
    #7 0x7fdadc80d6b9 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x76b9)

Thread T25 created by T0 here:
    #0 0x7fdadca823cd in pthread_create (/usr/lib/llvm-5.0/lib/clang/5.0.0/lib/linux/libclang_rt.asan-x86_64.so+0x5f3cd)
    #1 0x7fdac761c494 in rml::internal::thread_monitor::launch(void* (*)(void*), void*, unsigned long) /var/lib/jenkins/workspace/aten/src/ATen/../../../third_party/tbb/src/rml/include/../server/thread_monitor.h:221:12
    #2 0x7fdac7619e7a in tbb::internal::rml::private_worker::wake_or_launch() /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:300:21
    #3 0x7fdac7618c2d in tbb::internal::rml::private_server::wake_some(int) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:394:17
    #4 0x7fdac7618d91 in tbb::internal::rml::private_server::adjust_job_count_estimate(int) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/private_server.cpp:404:9
    #5 0x7fdac7612a5b in tbb::internal::market::adjust_demand(tbb::internal::arena&, int) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/market.cpp:678:16
    #6 0x7fdac76254dc in void tbb::internal::arena::advertise_new_work<(tbb::internal::arena::new_work_type)0>() /var/lib/jenkins/workspace/third_party/tbb/src/tbb/arena.h:451:24
    #7 0x7fdac761f396 in tbb::internal::generic_scheduler::local_spawn(tbb::task*, tbb::task*&) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/scheduler.cpp:648:15
    #8 0x7fdac761f9e8 in tbb::internal::generic_scheduler::spawn(tbb::task&, tbb::task*&) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/scheduler.cpp:674:34
    #9 0x7fdac7453c53 in tbb::interface5::internal::task_base::spawn(tbb::task&) /var/lib/jenkins/workspace/aten/src/ATen/../../../third_party/tbb/include/tbb/internal/../task.h:995:23
    #10 0x7fdac7453c53 in tbb::interface9::internal::start_for<tbb::blocked_range<long>, void at::parallel_for<void at::native::(anonymous namespace)::_parallel_for<at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1}>(long, long, bool, at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1})::{lambda(long, long)#1}>(long, long, long, at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1})::{lambda(tbb::blocked_range<long> const&)#1}, tbb::simple_partitioner const>::offer_work(tbb::split&) /var/lib/jenkins/workspace/aten/src/ATen/../../../third_party/tbb/include/tbb/parallel_for.h:122
    #11 0x7fdac7453c53 in void tbb::interface9::internal::simple_partition_type::execute<tbb::interface9::internal::start_for<tbb::blocked_range<long>, void at::parallel_for<void at::native::(anonymous namespace)::_parallel_for<at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1}>(long, long, bool, at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1})::{lambda(long, long)#1}>(long, long, long, at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1})::{lambda(tbb::blocked_range<long> const&)#1}, tbb::simple_partitioner const>, tbb::blocked_range<long> >(at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1}&, tbb::blocked_range<long>&) /var/lib/jenkins/workspace/aten/src/ATen/../../../third_party/tbb/include/tbb/partitioner.h:525
    #12 0x7fdac7453c53 in tbb::interface9::internal::start_for<tbb::blocked_range<long>, void at::parallel_for<void at::native::(anonymous namespace)::_parallel_for<at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1}>(long, long, bool, at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1})::{lambda(long, long)#1}>(long, long, long, at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1})::{lambda(tbb::blocked_range<long> const&)#1}, tbb::simple_partitioner const>::execute() /var/lib/jenkins/workspace/aten/src/ATen/../../../third_party/tbb/include/tbb/parallel_for.h:143
    #13 0x7fdac76334d5 in tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::local_wait_for_all(tbb::task&, tbb::task*) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/custom_scheduler.h:509:33
    #14 0x7fdac761f8b8 in tbb::internal::generic_scheduler::local_spawn_root_and_wait(tbb::task*, tbb::task*&) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/scheduler.cpp:670:5
    #15 0x7fdac761fa18 in tbb::internal::generic_scheduler::spawn_root_and_wait(tbb::task&, tbb::task*&) /var/lib/jenkins/workspace/third_party/tbb/src/tbb/scheduler.cpp:678:34
    #16 0x7fdac7452181 in tbb::task::spawn_root_and_wait(tbb::task&) /var/lib/jenkins/workspace/aten/src/ATen/../../../third_party/tbb/include/tbb/internal/../task.h:752:30
    #17 0x7fdac7452181 in tbb::interface9::internal::start_for<tbb::blocked_range<long>, void at::parallel_for<void at::native::(anonymous namespace)::_parallel_for<at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1}>(long, long, bool, at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1})::{lambda(long, long)#1}>(long, long, long, at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1})::{lambda(tbb::blocked_range<long> const&)#1}, tbb::simple_partitioner const>::run(void at::native::(anonymous namespace)::_parallel_for<at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1}>(long, long, bool, at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1})::{lambda(long, long)#1}, tbb::blocked_range<long> const&, {lambda(tbb::blocked_range<long> const&)#1}&) /var/lib/jenkins/workspace/aten/src/ATen/../../../third_party/tbb/include/tbb/parallel_for.h:96
    #18 0x7fdac7452181 in void tbb::parallel_for<tbb::blocked_range<long>, void at::parallel_for<void at::native::(anonymous namespace)::_parallel_for<at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1}>(long, long, bool, at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1})::{lambda(long, long)#1}>(long, long, long, at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1})::{lambda(tbb::blocked_range<long> const&)#1}>(at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1} const&, void at::parallel_for<void at::native::(anonymous namespace)::_parallel_for<at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1}>(long, long, bool, at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1})::{lambda(long, long)#1}>(long, long, long, at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1})::{lambda(tbb::blocked_range<long> const&)#1} const&, tbb::simple_partitioner const&) /var/lib/jenkins/workspace/aten/src/ATen/../../../third_party/tbb/include/tbb/parallel_for.h:209
    #19 0x7fdac7452181 in void at::parallel_for<void at::native::(anonymous namespace)::_parallel_for<at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1}>(long, long, bool, at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1})::{lambda(long, long)#1}>(long, long, long, at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1}) /var/lib/jenkins/workspace/aten/src/ATen/Parallel.h:39
    #20 0x7fdac7452181 in void at::native::(anonymous namespace)::_parallel_for<at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1}>(long, long, bool, at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long)::{lambda(long)#1}) /var/lib/jenkins/workspace/build/aten/src/ATen/native/cpu/ReduceOpsKernel.cpp.AVX2.cpp:23
    #21 0x7fdac7452181 in at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::reduce2d(double const*, double*, long, long, long) /var/lib/jenkins/workspace/build/aten/src/ATen/native/cpu/ReduceOpsKernel.cpp.AVX2.cpp:120
    #22 0x7fdac7452181 in at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::apply(at::Tensor&, at::Tensor const&, at::optional<long>)::{lambda(long)#1}::operator()(long) const /var/lib/jenkins/workspace/build/aten/src/ATen/native/cpu/ReduceOpsKernel.cpp.AVX2.cpp:74
    #23 0x7fdac7447a4a in void at::native::(anonymous namespace)::_parallel_for<at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::apply(at::Tensor&, at::Tensor const&, at::optional<long>)::{lambda(long)#1}>(long, long, bool, at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::apply(at::Tensor&, at::Tensor const&, at::optional<long>)::{lambda(long)#1}) /var/lib/jenkins/workspace/build/aten/src/ATen/native/cpu/ReduceOpsKernel.cpp.AVX2.cpp:31:7
    #24 0x7fdac7447a4a in at::native::(anonymous namespace)::Reduction<double, std::plus, 0>::apply(at::Tensor&, at::Tensor const&, at::optional<long>) /var/lib/jenkins/workspace/build/aten/src/ATen/native/cpu/ReduceOpsKernel.cpp.AVX2.cpp:70
    #25 0x7fdac7447a4a in at::native::(anonymous namespace)::sum_kernel_impl(at::Tensor&, at::Tensor const&, at::optional<long>)::$_0::operator()() const::{lambda()#3}::operator()() const /var/lib/jenkins/workspace/build/aten/src/ATen/native/cpu/ReduceOpsKernel.cpp.AVX2.cpp:147
    #26 0x7fdac7447a4a in at::native::(anonymous namespace)::sum_kernel_impl(at::Tensor&, at::Tensor const&, at::optional<long>)::$_0::operator()() const /var/lib/jenkins/workspace/build/aten/src/ATen/native/cpu/ReduceOpsKernel.cpp.AVX2.cpp:147
    #27 0x7fdac7447a4a in at::native::(anonymous namespace)::sum_kernel_impl(at::Tensor&, at::Tensor const&, at::optional<long>) /var/lib/jenkins/workspace/build/aten/src/ATen/native/cpu/ReduceOpsKernel.cpp.AVX2.cpp:147
    #28 0x7fdac5e9ccd0 in void at::native::DispatchStub<void (*)(at::Tensor&, at::Tensor const&, at::optional<long>)>::operator()<at::Tensor, at::Tensor, long>(at::Tensor, at::Tensor, long) /var/lib/jenkins/workspace/aten/src/ATen/native/cpu/CapabilityDispatch.h:43:5
    #29 0x7fdac5e9ccd0 in at::native::_sum_out_cpu(at::Tensor&, at::Tensor const&, long, bool) /var/lib/jenkins/workspace/aten/src/ATen/native/ReduceOps.cpp:182
    #30 0x7fdac5ea5cee in at::native::_sum_out(at::Tensor&, at::Tensor const&, long, bool) /var/lib/jenkins/workspace/aten/src/ATen/native/ReduceOps.cpp:336:12
    #31 0x7fdac5ea5cee in at::Tensor& at::native::reduce_multi_associative_out<&at::native::_sum, &at::native::_sum_out>(at::Tensor&, at::Tensor const&, at::ArrayRef<long>, bool) /var/lib/jenkins/workspace/aten/src/ATen/native/ReduceOps.cpp:310
    #32 0x7fdac5ea404c in at::native::_sum_out(at::Tensor&, at::Tensor const&, at::ArrayRef<long>, bool) /var/lib/jenkins/workspace/aten/src/ATen/native/ReduceOps.cpp:346:10
    #33 0x7fdac640fbab in at::Type::_sum_out(at::Tensor&, at::Tensor const&, at::ArrayRef<long>, bool) const /var/lib/jenkins/workspace/build/aten/src/ATen/Type.cpp:3979:13
    #34 0x7fdac5ea0892 in at::_sum_out(at::Tensor&, at::Tensor const&, at::ArrayRef<long>, bool) /var/lib/jenkins/workspace/build/aten/src/ATen/Functions.h:3736:29
    #35 0x7fdac5ea0892 in at::native::_sum(at::Tensor const&, long, bool) /var/lib/jenkins/workspace/aten/src/ATen/native/ReduceOps.cpp:242
    #36 0x7fdac5ea527d in at::Tensor at::native::reduce_multi_associative<&at::native::_sum>(at::Tensor const&, at::ArrayRef<long>, bool) /var/lib/jenkins/workspace/aten/src/ATen/native/ReduceOps.cpp:291:12
    #37 0x7fdac5ea4031 in at::native::_sum(at::Tensor const&, at::ArrayRef<long>, bool) /var/lib/jenkins/workspace/aten/src/ATen/native/ReduceOps.cpp:341:10
    #38 0x7fdac640fb0d in at::Type::_sum(at::Tensor const&, at::ArrayRef<long>, bool) const /var/lib/jenkins/workspace/build/aten/src/ATen/Type.cpp:3963:13
    #39 0x7fdacad1f38a in torch::autograd::VariableType::_sum(at::Tensor const&, at::ArrayRef<long>, bool) const /var/lib/jenkins/workspace/torch/csrc/autograd/generated/VariableType.cpp:27333:39
    #40 0x7fdac5ea03c4 in at::_sum(at::Tensor const&, at::ArrayRef<long>, bool) /var/lib/jenkins/workspace/build/aten/src/ATen/Functions.h:3724:29
    #41 0x7fdac5ea03c4 in at::native::sum(at::Tensor const&, at::ArrayRef<long>, bool, at::optional<at::ScalarType>) /var/lib/jenkins/workspace/aten/src/ATen/native/ReduceOps.cpp:224
    #42 0x7fdac5ea03c4 in at::native::sum(at::Tensor const&, at::ArrayRef<long>, bool) /var/lib/jenkins/workspace/aten/src/ATen/native/ReduceOps.cpp:232
    #43 0x7fdac640faad in at::Type::sum(at::Tensor const&, at::ArrayRef<long>, bool) const /var/lib/jenkins/workspace/build/aten/src/ATen/Type.cpp:3955:13
    #44 0x7fdacad1d9e9 in torch::autograd::VariableType::sum(at::Tensor const&, at::ArrayRef<long>, bool) const /var/lib/jenkins/workspace/torch/csrc/autograd/generated/VariableType.cpp:27299:23
    #45 0x7fdacb0a8d0f in at::Tensor::sum(at::ArrayRef<long>, bool) const /var/lib/jenkins/workspace/torch/lib/tmp_install/include/ATen/TensorMethods.h:1233:19
    #46 0x7fdacb4c915b in torch::autograd::dispatch_sum(at::Tensor&, at::ArrayRef<long>, bool) /var/lib/jenkins/workspace/torch/csrc/autograd/generated/python_variable_methods_dispatch.h:2360:15
    #47 0x7fdacb45d0b2 in torch::autograd::THPVariable_sum(_object*, _object*, _object*) /var/lib/jenkins/workspace/torch/csrc/autograd/generated/python_variable_methods.cpp:4412:17
    #48 0x555e86f7c753 in _PyCFunction_FastCallDict (/opt/conda/bin/python3.6+0x10f753)

SUMMARY: AddressSanitizer: heap-use-after-free /var/lib/jenkins/workspace/third_party/tbb/src/tbb/scheduler.h:571:26 in tbb::internal::generic_scheduler::is_worker() const
Shadow bytes around the buggy address:
  0x0c2c800b5f90: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c800b5fa0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c800b5fb0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c800b5fc0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c800b5fd0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x0c2c800b5fe0: fd fd fd fd fd fd fd fd fd[fd]fd fd fd fd fd fd
  0x0c2c800b5ff0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2c800b6000: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2c800b6010: fd fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c800b6020: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c800b6030: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==148==ABORTING
```

**Comment by cpuhrsch** [comment]
Will be revisited after reintroducing TBB.

**Pull request closed without merging**